File: P_1195_口袋的天空.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long

struct DSU { // 并查集
    vector<int> pre, siz;
    
    DSU() {}
    DSU(int n) {
        pre.resize(n + 1);
        iota(pre.begin(), pre.end(), 0);
        siz.assign(n + 1, 1);
    }
    
    int find(int x) {
        if (pre[x] == x) {
            return x;
        }
        return pre[x] = find(pre[x]);
    }
    
    bool same(int x, int y) {
        return find(x) == find(y);
    }
    
    bool merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) {
            return false;
        } else if (siz[x] < siz[y]) {
            swap(x, y);
        }
        siz[x] += siz[y];
        pre[y] = x;
        return true;
    }
    
    int size(int x) {
        return siz[find(x)];
    }
};

signed main() {
    int n, m, k, ans = 0, cnt = 0;
    cin >> n >> m >> k;
    DSU dsu(n);
    vector<tuple<int, int, int>> a;
    for (int i = 0; i < m; ++i) {
        int x, y, l;
        cin >> x >> y >> l;
        a.emplace_back(l, x, y);
    }
    sort(a.begin(), a.end());
    for (auto [l, x, y] : a) {
        if (dsu.merge(x, y)) {
            ans += l;
            if (++cnt == n - k) {
                cout << ans << endl;
                return 0;
            }
        }
    }
    cout << "No Answer" << endl;
}

File: P_1195_口袋的天空.py
class DSU: # 并查集
    def __init__(self, n):
        self.pre = list(range(n + 1))
        self.siz = [1 for _ in range(n + 1)]
    
    def find(self, x):
        x, xcopy = self.pre[x], x
        while x != self.pre[x]:
            x = self.pre[x]
        while xcopy != x:
            self.pre[xcopy], xcopy = x, self.pre[xcopy]
        return x
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def merge(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        self.siz[x] += self.siz[y]
        self.pre[y] = x
        return True
    
    def size(self, x):
        return self.siz[self.find(x)]

n, m, k = map(int, input().split())
ans, cnt = 0, 0
a = [list(map(int, input().split())) for _ in range(m)]
dsu = DSU(n)
a.sort(key = lambda x: x[2])
for x, y, l in a:
    if dsu.merge(x, y):
        ans += l
        cnt += 1
        if cnt == n - k:
            exit(print(ans))
print("No Answer")

